### ISSUE TITLE
Troubles with xgboost in the newest mlr version (parameter missing and gblinear)

### ISSUE THREAD

**Issue by rwarnung** [mlr-org/mlr#1504]
Hi,
today I updated mlr to the newest version and xgboost too. 

When I try 
`makeLearner("classif.xgboost")` then I get the error:

> Error in makeNumericLearnerParam(id = "missing", default = NULL, tunable = FALSE,  : 
>   unused argument (special.vals = list(NA, NA, NULL))

thus I wrote another learner that avoids this parameter. Let's call this learner xgboost2 (definition attached).
When I use this one using `makeDiscreteParam("booster", value = "gblinear", tunable = FALSE)`,  I get the error

```
 Error in predict.xgb.Booster(m, newdata = data.matrix(.newdata), ...) : 
  [17:05:01] amalgamation/../src/gbm/gblinear.cc:177: Check failed: (ntree_limit) == (0) GBLinear::Predict ntrees is only valid for gbtree predictor 
```

The latter seems to be an xgboost error. So I installed the newly updated version from github

`install.packages("xgboost", repos="http://dmlc.ml/drat/", type = "source")
`
but still get the error. I can not reproduce it using xgboost directly (see an example below). It seems to work. Can it be that mlr does not use the newly installed xboost version (the one installed from github)?

The  full test can be found below. All in all xgboost is really useful but there are some hiccups at the moment. Has anybody already found a way around this? 

Thank you!
Richard

PS:
The example:

```
library(mlr)
lrn = makeLearner("classif.xgboost2")

ps = makeParamSet(
  makeDiscreteParam("booster", value = "gblinear", tunable = FALSE),
  makeDiscreteParam("nrounds", values = 25*(1:40) ),
  makeDiscreteParam("max_depth", values = c(1L, 2L, 3L) ) ,
  makeDiscreteParam("eta", values = c(0.0001,0.05,0.0025) ),
  makeNumericParam("lambda",  lower = 0, upper = 200),
  makeNumericParam("lambda_bias",  lower = 0, upper = 200),
  makeNumericParam("alpha", lower = 0, upper = 200),
  makeDiscreteParam("colsample_bytree", values = (4:9)*0.1),
  makeDiscreteParam("subsample", values = (4:9)*0.1),
  makeDiscreteParam("min_child_weight", values = c(5L,10L,25L,50L)),
  makeDiscreteParam("early_stopping_rounds", values = 4, tunable = FALSE),
  makeDiscreteParam("verbose", values = 1, tunable = FALSE),
  makeDiscreteParam("eval_metric", values = c("logloss"), tunable = FALSE)
)
[inTrain,], one_of(modelvars) ),target = "target",positive = "pos", id = "Adv task")

res = tuneParams(lrn, task = sonar.task, resampling = makeResampleDesc("CV", iters = 3L),  control = makeTuneControlIrace(maxExperiments = 300L, show.irace.output = FALSE) , par.set = ps,measures = logloss, show.info = TRUE)
```
a similar call to xgboost on the other hand works:

```
require("xgboost")
data = as.matrix(getTaskData(sonar.task)[,1:60])
target = ifelse(getTaskData(sonar.task)$Class=="M",1,0)
res = xgboost(data = data,label=target, booster="gblinear", nrounds=82, max_depth=1, eta=0.05, lambda=154, lambda_bias=57.8, alpha=13.1, 
        colsample_bytree=0.5, subsample=0.8, min_child_weight=5, early_stopping_rounds=4, verbose=1,
        eval_metric="logloss")
```

**Comment by rwarnung** [comment]
The definition of xgboost2 was not attached. I paste it here:

```
makeRLearner.classif.xgboost2 = function() {
  makeRLearnerClassif(
    cl = "classif.xgboost2",
    package = "xgboost",
    par.set = makeParamSet(
      # we pass all of what goes in 'params' directly to ... of xgboost
      # makeUntypedLearnerParam(id = "params", default = list()),
      makeDiscreteLearnerParam(id = "booster", default = "gbtree", values = c("gbtree", "gblinear", "dart")),
      makeIntegerLearnerParam(id = "silent", default = 0L, tunable = FALSE),
      makeNumericLearnerParam(id = "eta", default = 0.3, lower = 0, upper = 1),
      makeNumericLearnerParam(id = "gamma", default = 0, lower = 0),
      makeIntegerLearnerParam(id = "max_depth", default = 6L, lower = 1L),
      makeNumericLearnerParam(id = "min_child_weight", default = 1, lower = 0),
      makeNumericLearnerParam(id = "subsample", default = 1, lower = 0, upper = 1),
      makeNumericLearnerParam(id = "colsample_bytree", default = 1, lower = 0, upper = 1),
      makeNumericLearnerParam(id = "colsample_bylevel", default = 1, lower = 0, upper = 1),
      makeIntegerLearnerParam(id = "num_parallel_tree", default = 1L, lower = 1L),
      makeNumericLearnerParam(id = "lambda", default = 0, lower = 0),
      makeNumericLearnerParam(id = "lambda_bias", default = 0, lower = 0),
      makeNumericLearnerParam(id = "alpha", default = 0, lower = 0),
      makeUntypedLearnerParam(id = "objective", default = "binary:logistic", tunable = FALSE),
      makeUntypedLearnerParam(id = "eval_metric", default = "error", tunable = FALSE),
      makeNumericLearnerParam(id = "base_score", default = 0.5, tunable = FALSE),
      makeNumericLearnerParam(id = "max_delta_step", lower = 0, default = 0),
      #makeNumericLearnerParam(id = "missing", default = NULL, tunable = FALSE),
      makeIntegerLearnerParam(id = "nthread", lower = 1L, tunable = FALSE),
      makeIntegerLearnerParam(id = "nrounds", default = 1L, lower = 1L),
      # FIXME nrounds seems to have no default in xgboost(), if it has 1, par.vals is redundant
      makeUntypedLearnerParam(id = "feval", default = NULL, tunable = FALSE),
      makeIntegerLearnerParam(id = "verbose", default = 1L, lower = 0L, upper = 2L, tunable = FALSE),
      makeIntegerLearnerParam(id = "print_every_n", default = 1L, lower = 1L, tunable = FALSE,
                              requires = quote(verbose == 1L)),
      makeIntegerLearnerParam(id = "early_stopping_rounds", default = 15L, lower = 1L, tunable = FALSE),
      makeLogicalLearnerParam(id = "maximize", default = FALSE, tunable = FALSE),
      makeDiscreteLearnerParam(id = "sample_type", default = "uniform", values = c("uniform", "weighted"), requires = quote(booster == "dart")),
      makeDiscreteLearnerParam(id = "normalize_type", default = "tree", values = c("tree", "forest"), requires = quote(booster == "dart")),
      makeNumericLearnerParam(id = "rate_drop", default = 0, lower = 0, upper = 1, requires = quote(booster == "dart")),
      makeNumericLearnerParam(id = "skip_drop", default = 0, lower = 0, upper = 1, requires = quote(booster == "dart"))
    ),
    par.vals = list(nrounds = 1L, verbose = 0L),
    properties = c("twoclass", "multiclass", "numerics", "factors", "prob", "weights", "missings", "featimp"),
    name = "eXtreme Gradient Boosting",
    short.name = "xgboost",
    note = "All settings are passed directly, rather than through `xgboost`'s `params` argument. `nrounds` has been set to `1` and `verbose` to `0` by default. `num_class` is set internally, so do not set this manually."
  )
}

#' @export
trainLearner.classif.xgboost2 = function(.learner, .task, .subset, .weights = NULL,  ...) {
  
  td = getTaskDescription(.task)
  parlist = list(...)
  parlist$data = data.matrix(getTaskData(.task, .subset, target.extra = TRUE)$data)
  parlist$label = match(as.character(getTaskData(.task, .subset, target.extra = TRUE)$target), td$class.levels) - 1
  nc = length(td$class.levels)
  
  if (is.null(parlist$objective))
    parlist$objective = ifelse(nc == 2L, "binary:logistic", "multi:softprob")
  
  if (.learner$predict.type == "prob" && parlist$objective == "multi:softmax")
    stop("objective = 'multi:softmax' does not work with predict.type = 'prob'")
  
  #if we use softprob or softmax as objective we have to add the number of classes 'num_class'
  if (parlist$objective %in% c("multi:softprob", "multi:softmax"))
    parlist$num_class = nc
  
  if (!is.null(.weights))
    parlist$data = xgboost::xgb.DMatrix(data = parlist$data, label = parlist$label, weight = .weights)
  
  do.call(xgboost::xgboost, parlist)
}

#' @export
predictLearner.classif.xgboost2 = function(.learner, .model, .newdata, ...) {
  td = .model$task.desc
  m = .model$learner.model
  cls = td$class.levels
  nc = length(cls)
  obj = .learner$par.vals$objective
  
  if (is.null(obj))
    .learner$par.vals$objective = ifelse(nc == 2L, "binary:logistic", "multi:softprob")
  
  p = predict(m, newdata = data.matrix(.newdata), ...)
  
  if (nc == 2L) { #binaryclass
    if (.learner$par.vals$objective == "multi:softprob") {
      y = matrix(p, nrow = length(p) / nc, ncol = nc, byrow = TRUE)
      colnames(y) = cls
    } else {
      y = matrix(0, ncol = 2, nrow = nrow(.newdata))
      colnames(y) = cls
      y[, 1L] = 1 - p
      y[, 2L] = p
    }
    if (.learner$predict.type == "prob") {
      return(y)
    } else {
      p = colnames(y)[max.col(y)]
      names(p) = NULL
      p = factor(p, levels = colnames(y))
      return(p)
    }
  } else { #multiclass
    if (.learner$par.vals$objective  == "multi:softmax") {
      return(factor(p, levels = cls)) #special handling for multi:softmax which directly predicts class levels
    } else {
      p = matrix(p, nrow = length(p) / nc, ncol = nc, byrow = TRUE)
      colnames(p) = cls
      if (.learner$predict.type == "prob") {
        return(p)
      } else {
        ind = max.col(p)
        cns = colnames(p)
        return(factor(cns[ind], levels = cns))
      }
    }
  }
}

#' @export
getFeatureImportanceLearner.classif.xgboost2 = function(.learner, .model, ...) {
  mod = getLearnerModel(.model)
  imp = xgboost::xgb.importance(feature_names = .model$features,
                                model = mod, ...)
  
  fiv = imp$Gain
  setNames(fiv, imp$Feature)
}

```

**Comment by schiffner** [comment]
Hi,

regarding the first error (`unused argument (special.vals = list(NA, NA, NULL))`): Please try updating ParamHelpers to the CRAN version (1.10).

@everyone: We require only ParamHelpers (≥ 1.8) in the mlr DESCRIPTION. I can set it to 1.10 (or is 1.9 sufficient?)

**Comment by berndbischl** [comment]
i guess i forgot to update the depency in the DESC. will do that now

@rwarnung simply update Paramhelpers from cran. that must resolve this

reopen if not

**Comment by berndbischl** [comment]
wait. 

ok regarding the 2nd error:

can you please tell me if the newest versions from cran for
- mlr
- PH
- xgboost 
work for you?

and whether there is any problem remaining now?

**Comment by schiffner** [comment]
I think the error is in xgboost.
With R 3.3.2, mlr 2.10, ParamHelpers 1.10, xgboost 0.6-4 I get:

```
lrn = makeLearner("classif.xgboost", booster="gblinear", nrounds=82, max_depth=1, eta=0.05, lambda=154, lambda_bias=57.8, alpha=13.1, 
	colsample_bytree=0.5, subsample=0.8, min_child_weight=5, early_stopping_rounds=4, verbose=1,
	eval_metric="logloss")
m = train(lrn, sonar.task)
p = predict(m, sonar.task)
Error in predict.xgb.Booster(m, newdata = data.matrix(.newdata), ...) : 
  [18:19:50] amalgamation/../src/gbm/gblinear.cc:179: Check failed: (ntree_limit) == (0) GBLinear::Predict ntrees is only valid for gbtree predictor
```

but also

```
library(xgboost)
data = as.matrix(getTaskData(sonar.task)[,1:60])
target = ifelse(getTaskData(sonar.task)$Class=="M",1,0)
res = xgboost(data = data,label=target, booster="gblinear", nrounds=82, max_depth=1, eta=0.05, lambda=154, lambda_bias=57.8, alpha=13.1, 
         colsample_bytree=0.5, subsample=0.8, min_child_weight=5, early_stopping_rounds=4, verbose=1,
         eval_metric="logloss")
predict(res, data)
Error in predict.xgb.Booster(res, data) : 
  [18:48:10] amalgamation/../src/gbm/gblinear.cc:179: Check failed: (ntree_limit) == (0) GBLinear::Predict ntrees is only valid for gbtree predictor

traceback()
3: .Call("XGBoosterPredict_R", object$handle, newdata, option[1], 
       as.integer(ntreelimit), PACKAGE = "xgboost")
2: predict.xgb.Booster(res, data)
1: predict(res, data)
```

**Comment by berndbischl** [comment]
confirmed.  minimal example seems to be (with recent cran versions for all packages)

```
lrn = makeLearner("classif.xgboost", booster="gblinear", early_stopping_rounds=4, verbose=1)
m = train(lrn, sonar.task)
p = predict(m, sonar.task)
```

@rwarnung 
thats not on our side, please report at xgboost

**Comment by rwarnung** [comment]
@schiffner and @berndbischl thank you for your quick answers. Updating PH was the key for the first problem! I realize (and thought so) that the second was on xgboost's side. Thanks for your efforts and for the great work on mlr.

**Comment by rwarnung** [comment]
I opened up an issue at xgboost: https://github.com/dmlc/xgboost/issues/2029

**Comment by khotilov** [comment]
The `Check failed: (ntree_limit) == (0)` error with gblinear was fixed in dmlc/xgboost#1929. However, it was merged on Jan6, a day after the most recent CRAN release. The next release will have the fix, and meanwhile you may install the development xgboost as described in https://xgboost.readthedocs.io/en/latest/build.html#installing-the-development-version

**Comment by edvbb** [comment]
Sorry
may be not in this topic, what is the analog of the parameter "scale_pos_weight" of xgboost in the package mlr?
